### PR TITLE
Solving 4498 - Poll newly added blockchain filters after first alarm task run

### LINKED ISSUE #4498: Raiden doesn't replay all past state changes before starting the Raiden Service
## Problem Definition
Raiden client doesn't replay all state changes before starting the Raiden Service.
(In my case that led to a crash of the echonode).

[raiden.log](https://github.com/raiden-network/raiden/files/3452929/raiden.log)

`2019-07-31 15:44:39.172715 [debug    ] Raiden Service started         [raiden.raiden_service] node=0xdD84b4E3B4Fb3b7AD427Fa0A0FF2009ca4B363da`
Happens way before replaying state changes from the past
`2019-07-31 15:44:57.085114 [debug    ] State changes                  [raiden.raiden_service] greenlet_name=AlarmTask._run node:0xdD84b4E3B4Fb3b7AD427Fa0A0FF2009ca4B363da node=0xdD84b4E3B4Fb3b7AD427Fa0A0FF2009ca4B363da state_changes=['{"block_number": "4830525", "gas_limit": "7073246", ... `

## Expectation
I'd expect the raiden client to replay all state changes from the past before starting the raiden service. 

## Reproduce
Starting Raiden with new datadir on any? (tested with goerli and rinkeby) network.

### System Description
Raiden version `v0.100.5.dev0`.
Used bbot internal ethnodes.

### PR DESCRIPTION
**Pull request by LefterisJP**
Fix #4498

### CODE REVIEW

File: raiden/raiden_service.py [hunk 1]
```diff
@@ -539,6 +529,33 @@ def _start_transport(self, chain_state: ChainState) -> None:
...
+    def _prepare_and_execute_alarm_first_run(self, from_block) -> None:
+        """Prepares the alarm task callback and executes its first run
+
+        Complete the first_run of the alarm task and synchronize with the
+        blockchain since the last run.
+
+         Notes about setup order:
+         - The filters must be polled after the node state has been primed,
+           otherwise the state changes won't have effect.
+         - The alarm must complete its first run before the transport is started,
```
Comment (hackaugusto) [inline raiden/raiden_service.py]: could you please add asserts for these two comments? Like so:

https://github.com/raiden-network/raiden/blob/c00eda0a89d2ff76b36d48b1f6da72f90fb9c422/raiden/raiden_service.py#L860-L861

File: raiden/raiden_service.py [hunk 2]
```diff
@@ -415,17 +415,7 @@ def start(self) -> None:
...
-        # Notes about setup order:
-        # - The filters must be polled after the node state has been primed,
-        # otherwise the state changes won't have effect.
-        # - The alarm must complete its first run before the transport is started,
-        #   to reject messages for closed/settled channels.
-        self.alarm.register_callback(self._callback_new_block)
-        self.alarm.first_run(last_log_block_number)
+        self._prepare_and_execute_alarm_first_run(from_block=last_log_block_number)
 
         chain_state = views.state_from_raiden(self)
```
Comment (hackaugusto) [inline raiden/raiden_service.py]: Please, add a comment here saying that this must be done after the alarm task ran and the wal was restored..

File: raiden/tests/integration/fixtures/raiden_network.py [hunk 3]
```diff
@@ -200,15 +201,16 @@ def raiden_network(
     confirmed_block = raiden_apps[0].raiden.confirmation_blocks + 1
     blockchain_services.deploy_service.wait_until_block(target_block_number=confirmed_block)
 
-    parallel_start_apps(raiden_apps)
+    if start_raiden_apps:
+        parallel_start_apps(raiden_apps)
```
Comment (hackaugusto) [inline raiden/tests/integration/fixtures/raiden_network.py]: Humm, so it looks that we don't need the apps to be started to run `create_all_channels_for_network`. Could you please merge this `if start_raiden_apps` with the other one bellow?

File: raiden/tests/integration/fixtures/raiden_network.py [hunk 4]
```diff
@@ -219,21 +221,29 @@ def raiden_network(
...
+            )
+
+        # Force blocknumber update
+        exception = RuntimeError("Alarm failed to start and set up start_block correctly")
+
+        with gevent.Timeout(seconds=5, exception=exception):
+            wait_for_alarm_start(raiden_apps)
+    else:
+        # simply wait for another confirmation blocks number since we can't
+        # confirm channels have been made through the raiden apps that are not started
```
Comment (hackaugusto) [inline raiden/tests/integration/fixtures/raiden_network.py]: Why is this necessary?

File: raiden/tests/integration/test_raidenservice.py [hunk 5]
```diff
@@ -162,3 +163,42 @@ def num_matching_queued_messages(room: str, message_type: Type) -> int:
...
+    original_first_run = app0.raiden._prepare_and_execute_alarm_first_run
+
+    def first_run_with_check(from_block):
+        """
+        This function simply enhances the alarm task first run
+
+        The enhanced version has a check for channels being available right after
+        the first run of the alarm task
+        """
+        original_first_run(from_block)
```
Comment (hackaugusto) [inline raiden/tests/integration/test_raidenservice.py]: This has race conditions, since `app0.start` will do the complete app initialization, the alarm task will be executed, and the channels may get pickup because of the alarm task and not the `_prepare_and_execute_alarm_first_run`.

Another solution is to just not monkeypatch it, and call the initialization methods up to the `_prepare_and_execute_alarm_first_run`, then assert on the expected result. Since the alarm task is not initialized then we know for sure there are not races with it.

File: raiden/tests/integration/fixtures/raiden_network.py (hunk 4 above)
Comment (LefterisJP) [reply raiden/tests/integration/fixtures/raiden_network.py]: Because there is no other way to wait until the channels ae created a sis done above when apps are actually started since the raiden apps are empty

File: raiden/tests/integration/test_raidenservice.py (hunk 5 above)
Comment (LefterisJP) [reply raiden/tests/integration/test_raidenservice.py]: Hmmm it is monkey patched. If it's not picked up by the first run of the alarm task this function should fail.

**Comment by LefterisJP** [comment]
> I don't see how this is fixing the problem, could you elaborate?

As I mentioned in the issue the problem with this bug is that in `after_new_token_network_create_filter()` the new blockchain filters are added but since it's the end of the alarm task they are not queried.

Instead they get queried later when new blocks come in.

This patch addresses this by querying the newly created filters again after the end of the alarm task to make sure all blockchain events are queries and generated.

You can see that this works by checking out the test I added. If you remove the blockchain events querying after the first run the test fails.

File: raiden/tests/integration/fixtures/raiden_network.py (hunk 4 above)
Comment (hackaugusto) [reply raiden/tests/integration/fixtures/raiden_network.py]: > Because there is no other way to wait until the channels ae created

The proxies will poll on the transaction hash and wait for the transaction to be confirmed. Is there anything else that has to be done?

File: raiden/tests/integration/test_raidenservice.py (hunk 5 above)
Comment (hackaugusto) [reply raiden/tests/integration/test_raidenservice.py]: > Hmmm it is monkey patched

`_prepare_and_execute_alarm_first_run` is, [`_start_alarm_task`](https://github.com/raiden-network/raiden/blob/1c1f5dba1a3aabdc87cbdfcb9c72ac0443f02aa2/raiden/raiden_service.py#L448) is not, and it will run, there is no synchronization among that method and the test, so there is a race condition.

Please avoid monkey patching.

File: raiden/tests/utils/network.py [hunk 6]
```diff
@@ -133,36 +133,39 @@ def payment_channel_open_and_deposit(
...
-        token_network_address=token_network_proxy.address,
-        channel_identifier=channel_identifier,
-    )
 
-    for app in [app0, app1]:
-        # Use each app's own chain because of the private key / local signing
-        token = app.raiden.chain.token(token_address)
-        payment_channel_proxy = app.raiden.chain.payment_channel(
-            canonical_identifier=canonical_identifier
+    if deposit != 0:
```
Comment (hackaugusto) [inline raiden/tests/utils/network.py]: A value `0` is invalid. Please raise an exception instead.

File: raiden/tests/integration/fixtures/raiden_network.py (hunk 4 above)
Comment (LefterisJP) [reply raiden/tests/integration/fixtures/raiden_network.py]: there are no proxes as raiden has not yet started.

File: raiden/tests/integration/test_raidenservice.py (hunk 5 above)
Comment (hackaugusto) [reply raiden/tests/integration/test_raidenservice.py]: okay, you're right, the monkeypatched function will run before the alarm task is started.

File: raiden/tests/integration/fixtures/raiden_network.py (hunk 3 above)
Comment (LefterisJP) [reply raiden/tests/integration/fixtures/raiden_network.py]: unfortunately we do:


https://circleci.com/gh/raiden-network/raiden/70268#tests/containers/3

File: raiden/tests/utils/network.py (hunk 6 above)
Comment (LefterisJP) [reply raiden/tests/utils/network.py]: as discussed since this is in test utils its fine

File: raiden/raiden_service.py [hunk 7]
```diff
@@ -539,6 +530,35 @@ def _start_transport(self, chain_state: ChainState) -> None:
...
+        Complete the first_run of the alarm task and synchronize with the
+        blockchain since the last run.
+
+         Notes about setup order:
+         - The filters must be polled after the node state has been primed,
+           otherwise the state changes won't have effect.
+         - The alarm must complete its first run before the transport is started,
+           to reject messages for closed/settled channels.
+        """
+        assert not self.transport, f"Transport is running. node:{self!r}"
```
Comment (hackaugusto) [inline raiden/raiden_service.py]: So, the asserts I had in mind were the two from the docstring:

- the node state has been primed
- before the transport is started

File: raiden/raiden_service.py [hunk 8]
```diff
@@ -539,6 +530,35 @@ def _start_transport(self, chain_state: ChainState) -> None:
...
+
+        self.alarm.register_callback(self._callback_new_block)
+        self.alarm.first_run(from_block)
+        # The first run of the alarm task processes some state changes and may add
+        # new token network event filters when this is the first time Raiden runs.
+        # Here we poll for any new events that may exist after the addition of
+        # those event filters.
+        latest_block_num = self.chain.get_block(block_identifier="latest")["number"]
```
Comment (hackaugusto) [inline raiden/raiden_service.py]: I believe this is polling unconfirmed data.

File: raiden/raiden_service.py [hunk 9]
```diff
@@ -539,6 +530,35 @@ def _start_transport(self, chain_state: ChainState) -> None:
...
+        blockchain_events = self.blockchain_events.poll_blockchain_events(latest_block_num)
+
+        state_changes = []
+        for event in blockchain_events:
+            state_changes.extend(blockchainevent_to_statechange(self, event, from_block))
+        if state_changes:
```
Comment (hackaugusto) [inline raiden/raiden_service.py]: Perhaps we should move this to `handle_and_track_state_changes`, then no caller has to do this check.

File: raiden/raiden_service.py (hunk 7 above)
Comment (LefterisJP) [reply raiden/raiden_service.py]: Hmm I put the asserts you asked in your last [comment](https://github.com/raiden-network/raiden/pull/4550#discussion_r311620861).

Apart from the second one since the alarm task needs to be primed after the first run.

> the node state has been primed

What do you mean by that? That the DB/wal is initialized?

File: raiden/raiden_service.py (hunk 8 above)
Comment (LefterisJP) [reply raiden/raiden_service.py]: you are right. Will fix this.

File: raiden/raiden_service.py (hunk 9 above)
Comment (LefterisJP) [reply raiden/raiden_service.py]: Good point. Also thought this may be a  good change.

File: raiden/raiden_service.py (hunk 8 above)
Comment (LefterisJP) [reply raiden/raiden_service.py]: Never mind I was wrong. The confirmed block number is the **from_block** for any blockchain queries. Not the **to_block**

File: raiden/raiden_service.py [hunk 10]
```diff
@@ -539,6 +530,43 @@ def _start_transport(self, chain_state: ChainState) -> None:
...
+        self.alarm.first_run(last_log_block)
+        # The first run of the alarm task processes some state changes and may add
+        # new token network event filters when this is the first time Raiden runs.
+        # Here we poll for any new events that may exist after the addition of
+        # those event filters.
+        latest_block_num = self.chain.get_block(block_identifier="latest")["number"]
+        latest_confirmed_block_num = max(
+            GENESIS_BLOCK_NUMBER, latest_block_num - self.confirmation_blocks
+        )
+        latest_confirmed_block_num += 1
```
Comment (hackaugusto) [inline raiden/raiden_service.py]: Humm, this is getting one unconfirmed block, it shouldn't be necessary

Comment (LefterisJP) [reply raiden/raiden_service.py]: hahaha wtf? How did this get in there? Thanks for seeing it.
